### ISSUE TITLE
Type Providers still don't compile in 15.7.2

### ISSUE THREAD

**Issue by VTJDailey** [Microsoft/visualfsharp#4972]
This is a repost of closed issue #4795 (I couldn't figure out how to post on that, or request it to be re-opened so I'm opening a new issue here)

When compiling a project that includes

    type MyService = Microsoft.FSharp.Data.TypeProviders.WsdlService<"theurl">

I get a compiler error from FSC: Could not find the file E:\path\tmpA17F.dll

comments in issue #4795 indicated that a fix for this was expected to be in the 15.7.2 release.  I'm not sure if this is an additional problem separate from what was fixed, or if the fix just didn't make it in time.

**Comment by VTJDailey** [comment]
hmm....I've already started reinstalling the older version of visual studio.  :(

I didn't know anything about the different source DLLs for type providers, so I'll try updating that first and confirm my code is still working on the old version.  If I don't have any problems there, I'll upgrade to the latest again and see how it goes.  It'll probably be tomorrow before I have a chance to do that and confirm for you though.

**Comment by VTJDailey** [comment]
OK, I was able to get right on this.

Upgrading back to 15.7.2 and adding the nuget package and referencing FSharp.Data.TypeProviders.WsdlService did not solve the problem, I'm still having the same problem.

I did a minimal project with the following steps:
* New F# console project
* Add FSharp.Data.TypeProviders (5.0.0.2)
* Add type definition in Program.fs

I found a public wsdl from amazon to replace my internal wsdl, and I still get the same error.
![image](https://user-images.githubusercontent.com/4733235/40379669-c016adba-5dc4-11e8-9908-859d6a7deb5a.png)

Zipped solution is here:
[Test.zip](https://github.com/Microsoft/visualfsharp/files/2027859/Test.zip)

**Comment by cartermp** [comment]
@KevinRansom can you take a look?

**Comment by KevinRansom** [comment]
On it

**Comment by VTJDailey** [comment]
Nope, only official releases.

The only oddity would be the downgrading -- I upgraded to 15.7.0, downgraded to 15.6.7 using the Microsoft instructions at https://docs.microsoft.com/en-us/visualstudio/productinfo/installing-an-earlier-release-of-vs2017, then upgraded back to 15.7.2.  I never touched 15.7.1.

**Comment by VTJDailey** [comment]
Hi @cartermp, here's what I've got.  I'm not sure exactly what you're looking to see but I can let you know the contents of any of these folders.  Also note that the parent folder on my machine is 'F#' not 'FSharp'.  I don't know if that's important.

![image](https://user-images.githubusercontent.com/4733235/40388398-6c3fc7bc-5ddd-11e8-9852-4682a9d51277.png)

**Comment by KevinRansom** [comment]
Still Looking

**Comment by KevinRansom** [comment]
@VTJDailey 

oh my goodness in the words of ancient treasure maps "Here there be Dragons".

Since I started looking at this, there seem to be at least 3 issues.

1.   The WSDL type provider exhibits the 'Could not find file 'C:\temp\wsdlrepro\ConsoleApplication1

````
\tmp4CB6.dll'.' behaviour
1>------ Rebuild All started: Project: ConsoleApplication1, Configuration: Debug Any CPU ------
1>FSC: error FS0193: Could not find file 'C:\temp\wsdlrepro\ConsoleApplication1\tmp4CB6.dll'.
1>Done building project "ConsoleApplication1.fsproj" -- FAILED.
1>
1>Build FAILED.
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
````


2.  The open source FSharp.Data.TypeProviders project fails with this:

````
1>------ Rebuild All started: Project: ConsoleApplication1, Configuration: Debug Any CPU ------
1>FSC: error FS3031: The type provider 'C:\temp\wsdlrepro\packages\FSharp.Data.TypeProviders.5.0.0.2\lib\net40\FSharp.Data.TypeProviders.dll' reported an error: Assembly attribute 'TypeProviderAssemblyAttribute' refers to a designer assembly 'C:\temp\wsdlrepro\packages\FSharp.Data.TypeProviders.5.0.0.2\lib\net40\FSharp.Data.TypeProviders.dll' which cannot be loaded or doesn't exist. Could not load file or assembly 'file:///C:\temp\wsdlrepro\packages\FSharp.Data.TypeProviders.5.0.0.2\lib\net40\FSharp.Data.TypeProviders.dll' or one of its dependencies. Operation is not supported. (Exception from HRESULT: 0x80131515)
1>FSC: warning FS3005: Referenced assembly 'C:\temp\wsdlrepro\packages\FSharp.Data.TypeProviders.5.0.0.2\lib\net40\FSharp.Data.TypeProviders.dll' has assembly level attribute 'Microsoft.FSharp.Core.CompilerServices.TypeProviderAssemblyAttribute' but no public type provider classes were found
1>Done building project "ConsoleApplication1.fsproj" -- FAILED.
1>
````

3.  The inbox fsharp type provider relies on the presence of releases of the dotnet sdk that do not ship with VS.  And when it searches using windows sdk, it uses the wrong registry keys.  The OSS is similarly broken, although recognises more recent net sdks.

Inbox supported framworks:  note winsdk, doesnot have: **WinSDK-NetFx40Tools** in registry
````
  "Software\\Microsoft\\Microsoft SDKs\\NETFXSDK\\4.6\\WinSDK-NetFx40Tools"
  "Software\\Microsoft\\Microsoft SDKs\\Windows\\v8.1A\\WinSDK-NetFx40Tools"
  "Software\\Microsoft\\Microsoft SDKs\\Windows\\v8.0A\\WinSDK-NetFx40Tools"
  "Software\\Microsoft\\Microsoft SDKs\\Windows\\v7.1\\WinSDK-NetFx40Tools"
  "Software\\Microsoft\\Microsoft SDKs\\Windows\\v7.0A\\WinSDK-NetFx40Tools"
````

Registry with WinSDK 8.1 installed
==================================

**Comment by davidoptima** [comment]
I also have still have this problem after installing 15.7.2 on a project that uses `WsdlService` type provider.

``` shell
Build FAILED.

       "C:\dev\xxx\src\Cli\Cli.fsproj" (Clean;Build target) (1) ->
       "C:\dev\xxx\src\Repository\Repository.fsproj" (default target) (2:3) ->
       (CoreCompile target) ->
         FSC : error FS0193: Could not find file 'C:\dev\xxx\src\Repository\tmp15A2.dll'. [C:\dev\xxx\src\Repository\Repository.fsproj]

    0 Warning(s)
    1 Error(s)
```

**Comment by vasily-kirichenko** [comment]
I have the same problem on a compiler built from master today.

**Comment by vasily-kirichenko** [comment]
The latest FSharp.Compiler.Tools does not have this bug, BTW.

**Comment by VTJDailey** [comment]
@KevinRansom, soooooo, what you're saying then is that I should prepare to be on 15.6.7 for a long time then? 😟 

Honestly, the best practical option for me, (and perhaps others following this issue) would be if I could somehow extract the generated service client type to written code and recompile it directly into my project.  However I'm not aware of any ability to do that from a type provider..

**Comment by vasily-kirichenko** [comment]
@VTJDailey reference FSharp.Compiler.Tools nuget package and you will get a working compiler that stays the same whatever machine or IDE do you use or have installed.

**Comment by KevinRansom** [comment]
@VTJDailey --- I have some F# 4.5 stuff to do, then I am on it, so no it shouldn't be that long.

**Comment by VTJDailey** [comment]
Actually, Vasily's suggestion of using the FSharp.Compiler.Tools package seems to have worked.  Making that change, the project now compiles for me.  I'll confess I don't quite understand how that would solve it, if at least part of the problem are errors you've uncovered within the WSDL type provider code.  But at the very least this seems to be a viable workaround.

Is it typical for people to run the compiler from a nuget package like that?  I didn't even realize that was a possibility before today.  It _seems_ like it would be a good idea, decoupling it from the Visual Studio installation, I just wouldn't want that decision to come back and bite me a month down the line.

**Comment by KevinRansom** [comment]
@VTJDailey .  It uses an old version of the compiler.

**Comment by VTJDailey** [comment]
Meaning a version <= the version from Visual Studo 15.6.7?

Do you mean it's a bad idea in general, only useful for the time being for this workaround?  The current version was released less than a month ago so it doesn't seem wildly out of date.  Unlike the FSharp.Core nuget package which seems ancient.

Or are you saying it won't play well with the F# tooling inside visual studio (code completion, navigation, etc.) because it's not a matched version?

Or are you just noting as an FYI that it's a slightly older version of the compiler, but it's fine to keep using.

**Comment by vasily-kirichenko** [comment]
@VTJDailey We've used the nuget package at work for several months with zero problems. It _may_ result with a slightly slower compilation because nothing in the package is `ngen`ed (however, you can do it manually once).

**Comment by KevinRansom** [comment]
@VTJDailey 

The compilertools is a fine packaging of the F# compiler.  It is however, less up to date.  If it works for you then life is good, keep doing what works.  I am greatfull that you noticed and reported our issues with the wsdl typeprovider.  Which is at least one benefit of staying up to date, along with the pain, we find out the issues quicker.

We release previews, early, but folk more often than not, wait until the RTM is released and then find the issues.  A change coming in 15.8 will make previews of F# much nicer, since we won't overwrite the global compiler.

**Comment by VTJDailey** [comment]
@KevinRansom Thank you for the info!  And good luck and what looks like a rough set of problems to address.  I'm sure it's very difficult to get any pre-release feedback.  I can't even imagine who is installing and making serious use of RTM releases outside of those doing active development on the visual studio environment, or maybe a handful of VS extension developers.

I know this thread is getting a little off course, but is there any particular reason the nuget package is behind the released visual studio version?  It sure is a whole lot easier to roll back to an earlier nuget package than to downgrade visual studio!

**Comment by cartermp** [comment]
FYI folks #5036 is addressing this. It appears that this set of type providers was utilizing _both_ native resources (`SqlDataConnection`) and managed resources (`WsdlService`). #4795 fixed the native resources issue, which we believed was the only regression at the time. #5036 will address managed resources.

**Comment by dsyme** [comment]
I don't understand why our existing tests for the inbox FSharp.Data.TypeProviders didn't catch these. 

Perhaps they were all disabled when we moved it out-of-box to. NuGet. But they shouldn't have been, we lost important coverage there. We should have rejigged them to at least use the NuGet.

**Comment by KevinRansom** [comment]
@dsyme 

The FSharp.Data.TypeProviders were removed back in January 2016, PR: here:
 https://github.com/Microsoft/visualfsharp/commit/5d8129f374de3c20669a090984b6caea0e48cd18

That PR removed the FSharp.Data.TypeProviderTesting.  I added end to end type provider tests recently based on the fsproj/tp sdk testing to ensure that we get .NETSdk/Desktop TP coverage.

However, we don't have testing based on the OSS FSharp.Data.TypeProviders, or the inbox provider.
I will add an issue to enable that type of testing.

Kevin

**Comment by davidoptima** [comment]
@KevinRansom what version of VS will this fix be released?

**Comment by cartermp** [comment]
@davidoptima This will for sure be available in the VS 15.8 release (and the forthcoming Preview 3).

We'll have to see if it's possible for this to get into a servicing release for VS 15.7. Right now there's no guarantee.

**Comment by cartermp** [comment]
This is now fixed in the 15.7.4 servicing release: https://github.com/Microsoft/visualfsharp/releases/tag/Visual-Studio-2017-Version-15.7.4

It is also in the forthcoming Preview 3.

**Comment by KevinRansom** [comment]
VS2017.8 Preview 3 -- desktop compiler only,
VS2017.8 Preview 4. -- desktop and dotnet cli F# compiler.
